src/ham/hamd/hamd.cpp

```diff
@@ -365,6 +366,40 @@ ::DBus::Struct< bool, std::string > hamd_c::set_roles(const std::string& login,
     return ret;
 }
 
+/**
+ * @brief Create a new user or modify an existing user
+ *
+ * @param login     User's login name
+ * @param roles     List of roles
+ * @param hashed_pw Hashed password. Must follow usermod's --password
+ *                  syntax.
+ */
+::DBus::Struct< bool, std::string > hamd_c::usermod(const std::string                & login,
+                                                    const std::vector< std::string > & roles,
+                                                    const std::string                & hashed_pw)
+{
+    ::DBus::Struct< bool,       /* success */
+                    std::string /* errmsg */   > ret;
+    struct passwd * pwd = ::getpwnam(login.c_str());
```

Comment:
We can't use `getpwnam()` to check if the user exists because under certain conditions (like when RADIUS or TACACS+ is configured) the `getpwnam()` may auto-magically create the user. That's because `getpwnam()` relies on the underlying NSS infrastructure, and when RADIUS/TACACS+ is configured this NSS infrastructure defaults to creating users that are missing from `/etc/passwd`.

Instead, you need to use the API `fgetpwent()`, which does not rely on NSS, and iterate through the file `/etc/passwd`. In other words, you can define and use the following function instead of `getpwnam()`:
```
/**
 * @brief Scan "/etc/passwd" looking for user. If found, return a pointer
 *        to a "struct passwd" containing all the data related to user.
 *
 * @param fn E.g. /etc/passwd
 * @param user The user we're looking for
 *
 * @return If user found, return a pointer to a struct passwd.
 */
static struct passwd * fgetpwnam(const char * user)
{
    struct passwd * pwd = NULL;
    FILE          * f   = fopen("/etc/passwd", "re");
    if (f)
    {
        struct passwd * ent;
        while (NULL != (ent = fgetpwent(f)))
        {
            if (streq(ent->pw_name, user))
            {
                pwd = ent;
                break;
            }
        }
        fclose(f);
    }

    return pwd;
}
```